grouper/repositories/audit_log.py

```diff
@@ -0,0 +1,100 @@
+from datetime import datetime
+from typing import TYPE_CHECKING
+
+from sqlalchemy import desc
+
+from grouper.entities.audit_log_entry import AuditLogEntry
+from grouper.models.audit_log import AuditLog, AuditLogCategory
+from grouper.models.group import Group
+from grouper.models.permission import Permission
+from grouper.models.user import User
+from grouper.plugin import get_plugin_proxy
+
+if TYPE_CHECKING:
+    from grouper.models.base.session import Session
+    from grouper.usecases.authorization import Authorization
+    from typing import List, Optional
+
+
+class UnknownUserException(Exception):
+    """User involved in a logged action does not exist."""
+
+    pass
+
+
+class AuditLogRepository(object):
+    """SQL storage layer for audit log entries."""
+
+    def __init__(self, session):
+        # type: (Session) -> None
+        self.session = session
+
+    def get_entries_affecting_permission(self, permission):
+        # type: (str) -> List[AuditLogEntry]
+        permission_obj = Permission.get(self.session, name=permission)
+        if not permission_obj:
+            return []
+        results = (
+            self.session.query(AuditLog)
+            .filter(AuditLog.on_permission_id == permission_obj.id)
+            .order_by(desc(AuditLog.log_time))
+        )
+        return [self._to_audit_log_entry(e) for e in results]
+
+    def log(
+        self,
+        authorization,  # type: Authorization
+        action,  # type: str
+        description,  # type: str
+        on_user=None,  # type: Optional[str]
+        on_group=None,  # type: Optional[str]
+        on_permission=None,  # type: Optional[str]
+        category=AuditLogCategory.general,  # type: AuditLogCategory
+    ):
+        # type: (...) -> None
+        """Log an action to the audit log.
+
+        Arguments don't cover all use cases yet.  This method will be expanded as further use cases
+        are ported to this service.
+        """
+        actor = User.get(self.session, name=authorization.actor)
+        if not actor:
+            raise UnknownUserException("unknown actor {}".format(authorization.actor))
+
+        # We currently have no way to log audit log entries for objects that no longer exist.  This
+        # should eventually be fixed via a schema change to use strings for all fields of the audit
+        # log.  For now, we'll die with an exception.
+        user_obj = User.get(self.session, name=on_user)
```

Comment:
This should probably have `if on_user else None` at the end, right? And we probably just want to enforce one of `on_user`, `on_group` or `on_permission` is set?

Reply:
I'm going to refactor this to use some helper methods, since my attempts to hack around it just made it worse. New diff coming.

There are some audit actions that don't act on any of a user, group, or permission. For instance, starting a global audit has only an actor, action, and description, and none of the other fields.